flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferCompressor.java

```diff
@@ -107,15 +107,28 @@ private int compress(Buffer buffer) {
                 "Illegal reference count, buffer need to be released.");
 
         try {
+            int compressedLen;
             int length = buffer.getSize();
-            // compress the given buffer into the internal heap buffer
-            int compressedLen =
-                    blockCompressor.compress(
-                            buffer.getNioBuffer(0, length),
-                            0,
-                            length,
-                            internalBuffer.getNioBuffer(0, internalBuffer.capacity()),
-                            0);
+            // if buffer is in-heap, manipulate the underlying array directly.
+            if (!buffer.getMemorySegment().isOffHeap()) {
```

Comment:
Good suggestion. In addition, for internalBuffer, can we promote the heap array to class field named `internalBufferArray`, and directly use it instead of `internalBuffer.array()` for the `!buffer.getMemorySegment().isOffHeap()` branch to avoid `ensureAccessible` every time?

Reply:
I think that is a good suggestion.